ledger-service/http-json/src/itlib/scala/http/AbstractWebsocketServiceIntegrationTest.scala

```diff
@@ -916,6 +938,111 @@ abstract class AbstractWebsocketServiceIntegrationTest
       }
   }
 
+  "Per-query offsets should override offsets issued on a per-query basis" in withHttpService {
+    (uri, _, _) =>
+      val dslSyntax = Consume.syntax[JsValue]
+      import dslSyntax._
+      import spray.json._
+      def createIouCommand(currency: String): String =
+        s"""{
+           |  "templateId": "Iou:Iou",
+           |  "payload": {
+           |    "observers": [],
+           |    "issuer": "Alice",
+           |    "amount": "999.99",
+           |    "currency": "$currency",
+           |    "owner": "Alice"
+           |  }
+           |}""".stripMargin
+      def createIou(currency: String): Future[Assertion] =
+        HttpServiceTestFixture
+          .postJsonStringRequest(
+            uri.withPath(Uri.Path("/v1/create")),
+            createIouCommand(currency),
+            headersWithAuth,
+          )
+          .map(_._1 shouldBe a[StatusCodes.Success])
+      def contractsQuery(currency: String): String =
+        s"""{"templateIds":["Iou:Iou"], "query":{"currency":"$currency"}}"""
+      def contractsQueryWithOffset(offset: domain.Offset, currency: String): String =
+        s"""{"templateIds":["Iou:Iou"], "query":{"currency":"$currency"}, "query_offset":"${offset.unwrap}"}"""
+      def contracts(currency: String, offset: Option[domain.Offset]): String =
+        offset.fold(contractsQuery(currency))(contractsQueryWithOffset(_, currency))
+      def ledgerEnd(expectedContracts: Int): Future[domain.Offset] = {
+        def go(killSwitch: UniqueKillSwitch): Sink[JsValue, Future[domain.Offset]] =
+          Consume.interpret(
+            for {
+              _ <- readAcsN(expectedContracts)
+              ContractDelta(Vector(), _, Some(offset)) <- readOne
+              _ = killSwitch.shutdown()
+              _ <- drain
+            } yield offset
+          )
+        val (killSwitch, source) =
+          singleClientQueryStream(jwt, uri, """{"templateIds":["Iou:Iou"]}""")
+            .viaMat(KillSwitches.single)(Keep.right)
+            .preMaterialize()
+        source.via(parseResp).runWith(go(killSwitch))
+      }
+      def test(
+          end: domain.Offset,
+          eurFrom: Option[domain.Offset],
+          usdFrom: Option[domain.Offset],
+          expected: Map[String, Int],
+      ): Future[Assertion] = {
+        def go(killSwitch: UniqueKillSwitch): Sink[JsValue, Future[Assertion]] = {
+          Consume.interpret(
+            for {
+              contracts <- readAcsAt(end)
+              result = contracts
+                .map(_._2.asJsObject.fields("currency").asInstanceOf[JsString].value)
+                .groupBy(identity)
+                .view
+                .mapValues(_.size)
+                .toMap
+              ContractDelta(Vector(), _, _) <- readOne
+              _ = println("*" * 80)
+              _ = println(contracts)
+              _ = println(result)
+              _ = println("*" * 80)
```

Comment:
Debug prints, will be removed before moving this out of draft.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/9847/commits/6a832849983f8d68bdbc48b8ede4b12bd84f16be